"Auto-fill on page load" options

eliykat opened the pull request:
UPDATED March 2021

## Objective

This is a PR for the Community Forums thread: [Disable “Auto-fill On Page Load” for specific logins](https://community.bitwarden.com/t/disable-auto-fill-on-page-load-for-specific-logins/14981).

This allows users to control auto-fill on page load (AOPL) settings on a per-login basis. I have added a new dropdown on the login edit screen:

[screenshot - now out of date](https://user-images.githubusercontent.com/31796059/98606134-68e35d00-2332-11eb-934a-db7edfb5a36e.PNG)

The options are:
- Use Global Settings - use the existing "Auto-fill On Page Load" tickbox in the Options tab to determine whether to auto-fill (this is currently the default behaviour for all logins)
- Always - ignore the global setting and always auto-fill
- Never - ignore the global setting and never auto-fill

## Code changes

* new property added to relevant `login` models, called `autofillOnPageLoad`. Uses new enum to track the 3 possible values.
* `autofiller.ts` now fires `doFillIfNeeded()` on **all pages all the time** (previously it only fired if the global AOPL setting was enabled).
* the cipher cache in `cipherService` is used to filter logins with AOPL enabled

Other repos:
* jslib: https://github.com/bitwarden/jslib/pull/199
* server: https://github.com/bitwarden/server/pull/986
* desktop: to do
* mobile: to do
* web: to do

## Comments
I'm keeping this in draft as it still needs to be implemented for the other repos before being merged. However, I'd like someone to look over it first to make sure the autofill logic is fine.

My main concerns are:

1. "Autofill On Page Load" is marked as experimental, but this expands the feature and reduces the user's ability to opt out. e.g. we are now running some of this code every time a page loads, regardless of any user setting. Is this OK for an experimental feature? When do we mark it as no longer experimental?
2. Are there any performance concerns now that we're checking for matching logins every single time a page is loaded? (again, regardless of whether the global option is enabled or not) I'm a bit concerned that this may add unacceptable overhead on general browser use, but I don't know the best way to measure it.

A more minor question: If a user configures AOPL for a login and saves to the server, and then edits the login from an older client without these settings available, then the updated login will have AOPL set to `null` and it will lose the previously saved value. Is there any way to avoid this, or do we just expect users to have up to date clients across the board?

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/bitwarden/browser?pullRequest=1455) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/bitwarden/browser?pullRequest=1455) before we can accept your contribution.<br/>**3** out of **5** committers have signed the CLA.<br/><br/>:white_check_mark: eliykat<br/>:white_check_mark: xusoo<br/>:white_check_mark: addisonbeck<br/>:x: kspearrin<br/>:x: vincentsalucci<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/bitwarden/browser?pullRequest=1455) it.</sub>

eliykat commented:
I've reviewed this, rebased against master, and fixed some minor issues. I'm keeping it in draft but would like a review before I go and implement across all other clients. See my updated comments/questions above.

vincentsalucci commented:
## Questions

> "Autofill On Page Load" is marked as experimental, but this expands the feature and reduces the user's ability to opt out. e.g. we are now running some of this code every time a page loads, regardless of any user setting. Is this OK for an experimental feature? When do we mark it as no longer experimental?

Good question - Since this is going to start firing on every page load, we might want to remove the "experimental" tag OR change this approach to still allow users a way to fully opt-out by moving the individual cipher decision a level higher. Example: Instead of a simple toggle in the global settings (for that platform), the options for Auto-fill on page load are either Never, Always, Use vault item settings. It would default to never (as-is, today) and would still give us the ability to leave the experimental tag because we have a top-level conditional that can escape prematurely from /always/ running `doFillIfNeeded()`.  On the cipher we would then have a simple toggle for Auto-fill on load (that is only visible if `Use vault item settings` is selected in the platform's options, default to false (0)).

This would probably require some more plumbing, but it seems to me like a better UX with reduced overhead and full opt-out-ability.

> Are there any performance concerns now that we're checking for matching logins every single time a page is loaded? (again, regardless of whether the global option is enabled or not) I'm a bit concerned that this may add unacceptable overhead on general browser use, but I don't know the best way to measure it.

My guess is that this overhead /might/ start to pile up for consumers with very large vaults. Would definitely need some insight from @kspearrin (since he wrote the original implementation).


## UI Considerations

Since this is still considered experimental - I really don't think the option to adjust this should be in the middle of the cipher options - we should stick it all the way at the bottom and it's default should be to use the global setting (as pictured/described).


## Side notes

- The `Auto-fill on page load` option is only available in the `browser` extension's settings(options). There would probably need to be some helper text/documentation on how/when this applies if we add the UI for adjusting the auto-fill option on the individual cipher level for /all/ platforms.

eliykat commented:
@vincentsalucci 

> > "Autofill On Page Load" is marked as experimental, but this expands the feature and reduces the user's ability to opt out. e.g. we are now running some of this code every time a page loads, regardless of any user setting. Is this OK for an experimental feature? When do we mark it as no longer experimental?
> 
> Good question - Since this is going to start firing on every page load, we might want to remove the "experimental" tag OR change this approach to still allow users a way to fully opt-out by moving the individual cipher decision a level higher. Example: Instead of a simple toggle in the global settings (for that platform), the options for Auto-fill on page load are either Never, Always, Use vault item settings. It would default to never (as-is, today) and would still give us the ability to leave the experimental tag because we have a top-level conditional that can escape prematurely from /always/ running `doFillIfNeeded()`. On the cipher we would then have a simple toggle for Auto-fill on load (that is only visible if `Use vault item settings` is selected in the platform's options, default to false (0)).
> 
> This would probably require some more plumbing, but it seems to me like a better UX with reduced overhead and full opt-out-ability.

I like this solution for the reasons you've listed. It also means we can wait and see about the performance issues, because if there are any, they'll only affect a small subset of users who have opted in to this feature. It leaves everyone else unaffected.

However, I think there is a UX tradeoff: there's no easy way of excluding only a few items from autofill on page load. e.g. if you want to autofill all items on page load except for (say) your internet banking credentials, you have to (1) set the global option to "Use vault item settings" and then (2) enable aopl on all your other vault items individually. This is probably only a one time cost, but might not be feasible for users with hundreds of vault items.

What if we keep the current item-level options as is, but separate the global options into "Enable feature" and "default behaviour". e.g.

![Screen Shot 2021-03-29 at 11 08 28 am](https://user-images.githubusercontent.com/31796059/112775120-40f55b80-907f-11eb-8972-3b8bc264d680.png)

Rationale:
* Maintains current option to enable/disable feature globally, so we can keep it experimental and opt-in only (basically lifts this up to a higher level as you suggest)
* Allows user to replicate current behaviour: enable feature and set default behaviour to "Always"
* Allows user to enable aopl for all items but exclude a subset: enable feature, default setting to "Always", then set excluded items to "Never"
* Allows a user to enable aopl for only a subset of items: enable feature, default setting to "Never", then set the desired items to "Always"

Downside: potentially confusing? I'm not sure anymore 😅

vincentsalucci commented:
I do think its a little confusing - but could be cleared up by good verbiage/clear explanation on the cipher override. 👍

eliykat commented:
Alright, here's the revised UI for feedback before I wire up the logic. This is the best/clearest way I can think of laying it out (plus help documentation explaining the feature in more detail).

Options page:
* now styled as a sub-option
* called it the "default" setting (rather than "global"), as that communicates the idea of being able to override it
* provided more context in the drop-down wording, "Always auto-fill" and "Never auto-fill"

![Screen Shot 2021-04-05 at 12 17 05 pm](https://user-images.githubusercontent.com/31796059/113530323-110f0080-9609-11eb-97b6-9755af03a812.png)

Add/Edit item screen:
* added "(if enabled)" in the label (to indicate there is a higher-level option)
* provided more context in the drop-down wording, "Use default setting", "Always auto-fill this login", "Never auto-fill this login".
* removed warning label, as that appears with the top-level option, and so I don't think we need to repeat it on every single item page
* @vincentsalucci Fair point about wanting to stick the experimental setting down the bottom of the Edit Item page, but now that I've removed the warning label I think it sits OK under the "Item Information" heading - otherwise it doesn't fit with the logical layout of the rest of the page. If we reintroduced the warning label on this page, I'd probably agree with you though so that we don't clutter up the rest of it.

![Screen Shot 2021-04-05 at 12 16 16 pm](https://user-images.githubusercontent.com/31796059/113530359-300d9280-9609-11eb-8669-b6e2c02103e0.png)

eliykat commented:
This is now out of draft and ready for review to be merged.

I made some further small changes to the UI, latest screenshots below. I ditched the Always/Never wording because I thought it was confusing.

Options page:

![Screen Shot 2021-04-16 at 1 59 12 pm](https://user-images.githubusercontent.com/31796059/114969798-73df8200-9ebc-11eb-9dba-1341acd36e00.png)

Login item page:

![Screen Shot 2021-04-16 at 2 00 10 pm](https://user-images.githubusercontent.com/31796059/114969808-79d56300-9ebc-11eb-899c-0e99ad14c3df.png)

![Screen Shot 2021-04-16 at 2 00 20 pm](https://user-images.githubusercontent.com/31796059/114969814-7c37bd00-9ebc-11eb-864d-f4ac119cbb16.png)

cscharf commented on src/popup/vault/add-edit.component.html:
```diff
@@ -268,6 +268,16 @@
                 </a>
             </div>
         </div>
+        <div class="box" *ngIf="cipher.type === cipherType.Login">
```
I would assume this only be visible / configurable (outside of `default`) if the underlying `Enable auto-fill on Page Load` checkbox/setting is checked/enabled in the options... otherwise this will, IMHO, present a confusing paradigm to the user where they're configuring auto-fill on page load settings when it's not even enabled overall in their settings. I love the concept of JIT or "just enough" configurable with smart defaults, but let's leave this option off the screen if it's not even applicable to avoid confusion (I can see the CS tickets and GitHub issues now with, "but I enabled it for this item".... us: "did you enable it in your overall options?").

eliykat replied:
I agree this could be confusing... let's do CS a favour I guess ;)
